// @nativescript/secure-storage/secure-storage.android.js
import { SecureStorageCommon } from "./secure-storage.common";
import { Utils } from '@nativescript/core';
export class SecureStorage extends SecureStorageCommon {
    constructor() {
        super();
        this.hawk = com.orhanobut.hawk.Hawk.init(Utils.android.getApplicationContext()).build();
    }
    get(arg) {
        return new Promise((resolve, reject) => {
            resolve(com.orhanobut.hawk.Hawk.get(arg.key));
        });
    }
    getSync(arg) {
        return com.orhanobut.hawk.Hawk.get(arg.key);
    }
    set(arg) {
        return new Promise((resolve, reject) => {
            resolve(com.orhanobut.hawk.Hawk.put(arg.key, arg.value));
        });
    }
    setSync(arg) {
        return com.orhanobut.hawk.Hawk.put(arg.key, arg.value);
    }
    remove(arg) {
        return new Promise((resolve, reject) => {
            resolve(com.orhanobut.hawk.Hawk.delete(arg.key));
        });
    }
    removeSync(arg) {
        return com.orhanobut.hawk.Hawk.delete(arg.key);
    }
    removeAll(arg) {
        return new Promise((resolve, reject) => {
            resolve(com.orhanobut.hawk.Hawk.deleteAll());
        });
    }
    removeAllSync(arg) {
        return com.orhanobut.hawk.Hawk.deleteAll();
    }
}
