// @nativescript/datetimepicker/angular/esm2020/index.mjs
import { NgModule } from "@angular/core";
import { registerElement } from "@nativescript/angular";
import { DatePickerField, TimePickerField, DateTimePickerFields } from "@nativescript/datetimepicker";
import { DatePickerFieldDirective, TimePickerFieldDirective, DateTimePickerFieldsDirective, } from "./nativescript-datetimepicker.directives";
import { DatePickerValueAccessor, TimePickerValueAccessor, DateTimePickersValueAccessor, } from "./nativescript-datetimepicker.accessors";
import * as i0 from "@angular/core";
export { DatePickerFieldDirective, TimePickerFieldDirective, DateTimePickerFieldsDirective, } from "./nativescript-datetimepicker.directives";
export { DatePickerValueAccessor, TimePickerValueAccessor, DateTimePickersValueAccessor, } from "./nativescript-datetimepicker.accessors";
export class NativeScriptDateTimePickerModule {
}
NativeScriptDateTimePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NativeScriptDateTimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NativeScriptDateTimePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.4", ngImport: i0, type: NativeScriptDateTimePickerModule, declarations: [DatePickerFieldDirective,
        TimePickerFieldDirective,
        DateTimePickerFieldsDirective,
        DatePickerValueAccessor,
        TimePickerValueAccessor,
        DateTimePickersValueAccessor], exports: [DatePickerFieldDirective,
        TimePickerFieldDirective,
        DateTimePickerFieldsDirective,
        DatePickerValueAccessor,
        TimePickerValueAccessor,
        DateTimePickersValueAccessor] });
NativeScriptDateTimePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NativeScriptDateTimePickerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NativeScriptDateTimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DatePickerFieldDirective,
                        TimePickerFieldDirective,
                        DateTimePickerFieldsDirective,
                        DatePickerValueAccessor,
                        TimePickerValueAccessor,
                        DateTimePickersValueAccessor,
                    ],
                    exports: [
                        DatePickerFieldDirective,
                        TimePickerFieldDirective,
                        DateTimePickerFieldsDirective,
                        DatePickerValueAccessor,
                        TimePickerValueAccessor,
                        DateTimePickersValueAccessor,
                    ],
                }]
        }] });
registerElement("DatePickerField", () => DatePickerField);
registerElement("TimePickerField", () => TimePickerField);
registerElement("DateTimePickerFields", () => DateTimePickerFields);
