// @nativescript/datetimepicker/angular/esm2020/nativescript-datetimepicker.directives.mjs
import { Directive } from "@angular/core";
import * as i0 from "@angular/core";
export class DatePickerFieldDirective {
}
DatePickerFieldDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: DatePickerFieldDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
DatePickerFieldDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.4", type: DatePickerFieldDirective, selector: "DatePickerField", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: DatePickerFieldDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "DatePickerField"
                }]
        }] });
export class TimePickerFieldDirective {
}
TimePickerFieldDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: TimePickerFieldDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TimePickerFieldDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.4", type: TimePickerFieldDirective, selector: "TimePickerField", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: TimePickerFieldDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "TimePickerField"
                }]
        }] });
export class DateTimePickerFieldsDirective {
}
DateTimePickerFieldsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: DateTimePickerFieldsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
DateTimePickerFieldsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.4", type: DateTimePickerFieldsDirective, selector: "DateTimePickerFields", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: DateTimePickerFieldsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "DateTimePickerFields"
                }]
        }] });
