// nativescript-barcodescanner/barcodescanner-common.js
import { ContentView, Property, booleanConverter } from "@nativescript/core";
export const formatsProperty = new Property({
    name: "formats",
    defaultValue: null,
});
export const preferFrontCameraProperty = new Property({
    name: "preferFrontCamera",
    defaultValue: false,
    valueConverter: booleanConverter
});
export const beepOnScanProperty = new Property({
    name: "beepOnScan",
    defaultValue: true,
    valueConverter: booleanConverter
});
export const reportDuplicatesProperty = new Property({
    name: "reportDuplicates",
    defaultValue: false,
    valueConverter: booleanConverter
});
export const pauseProperty = new Property({
    name: "pause",
    defaultValue: false,
    valueConverter: booleanConverter
});
export class BarcodeScannerView extends ContentView {
    pauseScanning() {
    }
    resumeScanning() {
    }
    [formatsProperty.setNative](value) {
        this.formats = value;
    }
    [preferFrontCameraProperty.setNative](value) {
        this.preferFrontCamera = value;
    }
    [beepOnScanProperty.setNative](value) {
        this.beepOnScan = value;
    }
    [reportDuplicatesProperty.setNative](value) {
        this.reportDuplicates = value;
    }
    [pauseProperty.setNative](value) {
        this.pause = value;
        this.pause ? this.pauseScanning() : this.resumeScanning();
    }
}
BarcodeScannerView.scanResultEvent = "scanResult";
pauseProperty.register(BarcodeScannerView);
formatsProperty.register(BarcodeScannerView);
preferFrontCameraProperty.register(BarcodeScannerView);
beepOnScanProperty.register(BarcodeScannerView);
reportDuplicatesProperty.register(BarcodeScannerView);
